src/libsync/discovery.cpp

```diff
@@ -890,77 +890,105 @@ void ProcessDirectoryJob::processFileAnalyzeLocalInfo(
-            _discoveryData->_renamedItemsLocal.insert(originalPath, path._target);
-            item->_renameTarget = path._target;
-            path._server = adjustedOriginalPath;
-            item->_file = path._server;
-            path._original = originalPath;
-            item->_originalFile = path._original;
-            item->_modtime = base._modtime;
-            item->_inode = base._inode;
-            item->_instruction = CSYNC_INSTRUCTION_RENAME;
-            item->_direction = SyncFileItem::Up;
-            item->_fileId = base._fileId;
-            item->_remotePerm = base._remotePerm;
-            item->_etag = base._etag;
-            item->_type = base._type;
-
-            // Discard any download/dehydrate tags on the base file.
-            // They could be preserved and honored in a follow-up sync,
-            // but it complicates handling a lot and will happen rarely.
-            if (item->_type == ItemTypeVirtualFileDownload)
-                item->_type = ItemTypeVirtualFile;
-            if (item->_type == ItemTypeVirtualFileDehydration)
-                item->_type = ItemTypeFile;
+    // Check local permission if we are allowed to put move the file here
+    // Technically we should use the permissions from the server, but we'll assume it is the same
+    auto movePerms = checkMovePermissions(base._remotePerm, originalPath, item->isDirectory());
+    if (!movePerms.sourceOk || !movePerms.destinationOk) {
+        qCInfo(lcDisco) << "Move without permission to rename base file, source:"
+                        << movePerms.sourceOk << ", target:" << movePerms.destinationOk;
 
-            qCInfo(lcDisco) << "Rename detected (up) " << item->_file << " -> " << item->_renameTarget;
-        };
+        // If we can create the destination, do that.
+        // Permission errors on the destination will be handled by checkPermissions later.
+        postProcessLocalNew();
+        finalize();
+
+        // If the destination upload will work, we're fine with the source deletion.
+        // If the source deletion can't work, checkPermissions will error.
+        if (movePerms.destinationOk)
+            return;
+
+        // Here we know the new location can't be uploaded: must prevent the source delete.
+        // Two cases: either the source item was already processed or not.
+        auto wasDeletedOnClient = _discoveryData->findAndCancelDeletedJob(originalPath);
         if (wasDeletedOnClient.first) {
-            recurseQueryServer = wasDeletedOnClient.second == base._etag ? ParentNotChanged : NormalQuery;
-            processRename(path);
+            // More complicated. The REMOVE is canceled. Restore will happen next sync.
+            qCInfo(lcDisco) << "Undid remove instruction on source" << originalPath;
+            _discoveryData->_statedb->deleteFileRecord(originalPath, true);
+            _discoveryData->_anotherSyncNeeded = true;
```

Comment:
Is it not kind of simple to do that now?

We can just create a new item with CSYNC_INSTRUCTION_NEW and _isRestoration = true.
Or is there something more complicated i don't see?

Reply:
We may need to restore a whole subtree recursively.

Reply:
Good point. Although that shouldn't be too complicated to recurse. We already do it when we delete a subtree.